{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92dd50c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import itertools\n",
    "from Bio import SeqIO\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.lines import Line2D\n",
    "import seaborn as sns\n",
    "import ast"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b06b3b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#input data\n",
    "##list of GSE studies\n",
    "gse_id_list = ['GSE145926', 'GSE154244', 'GSE150316']\n",
    "\n",
    "##min for variant frequency\n",
    "min_VF = 0.001\n",
    "\n",
    "##mapping method\n",
    "method = 'BWA'\n",
    "\n",
    "##read genome SARS-CoV-2 file\n",
    "##to find position in genome: str(genome_dict['NC_045512.2'].seq)\n",
    "genome_file = open('SARS_CoV_2.fasta')\n",
    "genome_dict = SeqIO.to_dict(SeqIO.parse(genome_file, \"fasta\"))\n",
    "\n",
    "#hairpins coordinates in SARS-CoV-2 (positions start with 1)\n",
    "hairpin_SARS_coordinates = 'hairpin_list.txt'\n",
    "df_hairpin = pd.read_csv(hairpin_SARS_coordinates, skiprows=1, names=['Loop_length', 'Pos', 'Loop_sequence', 'Stem_length'])\n",
    "df_hairpin['end_pos'] = df_hairpin['Pos'].apply(lambda x: ast.literal_eval(x)[0][1]-1)\n",
    "\n",
    "##palette for pictures\n",
    "colors_list = ['#C6878F', '#81B29A']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39b21942",
   "metadata": {},
   "outputs": [],
   "source": [
    "def apobec_mutagenesis(vcf_file, ID, align_method, min_vf=min_VF, genome=str(genome_dict['NC_045512.2'].seq), hairpin_SARS_coordinates=df_hairpin, colors=colors_list):\n",
    "    \n",
    "    ##read vcf file\n",
    "    ##skip description in vcf file\n",
    "    with open(vcf_file, 'r') as f:\n",
    "        reader=f.readlines()\n",
    "        row = 0\n",
    "        while reader[row].startswith('##') == True:\n",
    "            row += 1\n",
    "    df_vcf = pd.read_csv(vcf_file, skiprows=row, sep='\\t')\n",
    "    df_vcf = df_vcf.dropna()\n",
    "    \n",
    "    #choose positions with C>T mutation\n",
    "    df_vcf_C_T = df_vcf[(df_vcf['REF']== 'C') & (df_vcf['ALT']== 'T')]\n",
    "    df_vcf_C_T.reset_index(inplace=True, drop=True)\n",
    "    if len(df_vcf_C_T) == 0:\n",
    "        return 'empty'\n",
    "    \n",
    "    #filter by min VF\n",
    "    FORMAT = df_vcf_C_T.iloc[0, -2].split(':')\n",
    "    VF_ind = FORMAT.index('VF')\n",
    "    df_vcf_C_T['VF'] = df_vcf_C_T.iloc[:,-1].apply(lambda x: float(x.split(':')[VF_ind]))\n",
    "    df_vcf_C_T = df_vcf_C_T[(df_vcf_C_T['VF']>=min_vf)]\n",
    "    \n",
    "    df_vcf_C_T['preceding_nucl'] = '-'\n",
    "    df_vcf_C_T['loop_position'] = '-'\n",
    "    df_vcf_C_T.reset_index(inplace=True, drop=True)\n",
    "    for row in range(0, len(df_vcf_C_T)):\n",
    "        mutation_pos = df_vcf_C_T.loc[row, 'POS'] - 1\n",
    "        #write letter before mutation C>T\n",
    "        df_vcf_C_T.at[row, 'preceding_nucl'] = genome[mutation_pos-1]\n",
    "        \n",
    "        #write if mutation is in a loop\n",
    "        end_loop_positions = hairpin_SARS_coordinates.end_pos.to_list()\n",
    "        if mutation_pos in end_loop_positions:\n",
    "            df_vcf_C_T.at[row, 'loop_position'] = '+'\n",
    "     \n",
    "    \n",
    "    #make plot\n",
    "    f, ax = plt.subplots(figsize=(5, 7))\n",
    "    a = Line2D([], [], color=colors[0], label='-', marker='D', linewidth=1)\n",
    "    b = Line2D([], [], color=colors[1], label='+', marker='D', linewidth=1)\n",
    "\n",
    "    sns.stripplot(data=df_vcf_C_T, x=\"preceding_nucl\", y=\"VF\", hue=\"loop_position\", marker='D', size=12, dodge=True,\n",
    "                  alpha=.2, linewidth=1, jitter=False, palette=colors, hue_order=['-', '+'], order=['A', 'T', 'C', 'G'])\n",
    "    ax.set(xlabel='Nucleotide before C')\n",
    "    ax.legend(handles=[a, b], title='Mutation is in a loop')\n",
    "    f.suptitle('C>T mutations in '+ID+'\\naligned using '+align_method)\n",
    "    plt.grid()\n",
    "    plt.tight_layout()\n",
    "    plt.close(f)\n",
    "    return(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53e9613b",
   "metadata": {},
   "outputs": [],
   "source": [
    "for gse_id in gse_id_list:\n",
    "    \n",
    "    ##get list of srr for gse_id\n",
    "    srr_acc_file_txt = \"SRR_Acc_List_\"+gse_id+\".txt\"\n",
    "    df_srr_acc = pd.read_csv(srr_acc_file_txt, names=['acc'])\n",
    "    srr_list = df_srr_acc.acc.to_list()\n",
    "    \n",
    "    ##path for directory with vcf files for gse_id\n",
    "    directory = 'bwa/vcf_'+gse_id\n",
    "    for srr_id in srr_list:\n",
    "        print(srr_id, end=', ')\n",
    "        vcf = directory + '/' + srr_id + '_sorted.vcf'\n",
    "        plot = apobec_mutagenesis(vcf, srr_id, method)\n",
    "        if plot == 'empty':\n",
    "            print('no data')\n",
    "        else:\n",
    "            plot.savefig(directory+'/pictures/'+srr_id+'_CT_mutations_'+method+'.png', dpi=800)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
