{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92dd50c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import itertools\n",
    "from Bio import SeqIO\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.lines import Line2D\n",
    "import seaborn as sns\n",
    "import ast"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b06b3b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#input data\n",
    "##list of GSE studies\n",
    "gse_id_list = ['GSE145926', 'GSE154244', 'GSE150316']\n",
    "\n",
    "##min for variant frequency\n",
    "min_VF = 0.001\n",
    "max_VF = 1\n",
    "##min coverage\n",
    "min_DP = 50\n",
    "\n",
    "#hairpins coordinates in SARS-CoV-2 (positions start with 1)\n",
    "hairpin_SARS_coordinates = 'hairpin_list.txt'\n",
    "df_hairpin = pd.read_csv(hairpin_SARS_coordinates, skiprows=1, names=['Loop_length', 'Pos', 'Loop_sequence', 'Stem_length'])\n",
    "df_hairpin['loop_start_pos'] = df_hairpin['Pos'].apply(lambda x: ast.literal_eval(x)[0][0]-1)\n",
    "df_hairpin['loop_end_pos'] = df_hairpin['Pos'].apply(lambda x: ast.literal_eval(x)[0][1]-1)\n",
    "start_loop_positions = df_hairpin.loop_start_pos.to_list()\n",
    "end_loop_positions = df_hairpin.loop_end_pos.to_list()\n",
    "list_loop_coordinates = [pd.Interval(left=start_loop_positions[i], right=end_loop_positions[i]) for i in range(0, len(start_loop_positions))]\n",
    "\n",
    "##palette for pictures\n",
    "colors_list = ['#C6878F', '#81B29A']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39b21942",
   "metadata": {},
   "outputs": [],
   "source": [
    "def apobec_mutagenesis(vcf_file, ID, align_method, min_vf=min_VF, max_vf=max_VF, min_dp=min_DP, hairpin_SARS_coordinates=list_loop_coordinates, colors=colors_list):\n",
    "    flag = False\n",
    "    ##read vcf file\n",
    "    ##skip description in vcf file\n",
    "    with open(vcf_file, 'r') as f:\n",
    "        reader=f.readlines()\n",
    "        row = 0\n",
    "        while reader[row].startswith('##') == True:\n",
    "            row += 1\n",
    "    df_vcf = pd.read_csv(vcf_file, skiprows=row, sep='\\t')\n",
    "    df_vcf.columns.values[-1] = \"FORMAT_numbers\"\n",
    "    df_vcf = df_vcf.dropna()\n",
    "    \n",
    "    #choose positions with C>T mutation\n",
    "    df_vcf_C_T = df_vcf[(df_vcf['REF']== 'C') & (df_vcf['ALT']== 'T')]\n",
    "    df_vcf_C_T.reset_index(inplace=True, drop=True)\n",
    "    if df_vcf_C_T.empty:\n",
    "        return 'empty'\n",
    "    \n",
    "    #filter by min and amx VF\n",
    "    FORMAT = df_vcf_C_T.iloc[0, -2].split(':')\n",
    "    VF_ind = FORMAT.index('VF')\n",
    "    df_vcf_C_T['VF'] = df_vcf_C_T.apply(lambda x: float(x['FORMAT_numbers'].split(':')[VF_ind]), axis=1)\n",
    "    df_vcf_C_T = df_vcf_C_T[(df_vcf_C_T['VF']>=min_vf)]\n",
    "    df_vcf_C_T = df_vcf_C_T[(df_vcf_C_T['VF']<max_vf)]\n",
    "    if df_vcf_C_T.empty:\n",
    "        return 'empty'\n",
    "    \n",
    "    #filter by min coverage\n",
    "    DP_ind = FORMAT.index('DP')\n",
    "    df_vcf_C_T['coverage'] = df_vcf_C_T.apply(lambda x: float(x['FORMAT_numbers'].split(':')[DP_ind]), axis=1)\n",
    "    df_vcf_C_T = df_vcf_C_T[(df_vcf_C_T['coverage']>=min_dp)]\n",
    "    if df_vcf_C_T.empty:\n",
    "        return 'empty'\n",
    "    \n",
    "    df_vcf_C_T['pos_in_loop'] = '-'\n",
    "    df_vcf_C_T.reset_index(inplace=True, drop=True)\n",
    "    \n",
    "    for row in range(0, len(df_vcf_C_T)):\n",
    "        mutation_pos = df_vcf_C_T.loc[row, 'POS'] - 1\n",
    "        #write if mutation is in a loop\n",
    "        for interval in hairpin_SARS_coordinates:\n",
    "            if mutation_pos in interval:\n",
    "                df_vcf_C_T.at[row, 'pos_in_loop'] = '+'\n",
    "                break\n",
    "    \n",
    "    #sorting by VF\n",
    "    df_vcf_C_T = df_vcf_C_T.sort_values(by='VF', ascending=False)\n",
    "    df_vcf_C_T.reset_index(inplace=True, drop=True)\n",
    "    df_vcf_C_T = df_vcf_C_T[['POS', 'VF', 'coverage', 'pos_in_loop']]\n",
    "    \n",
    "    #make plot\n",
    "    ##change figure size\n",
    "    if len(df_vcf_C_T) > 200:\n",
    "        \n",
    "        df_vcf_C_T_abbr = df_vcf_C_T[df_vcf_C_T['VF']>=0.002]\n",
    "        df_vcf_C_T_abbr.reset_index(inplace=True, drop=True)\n",
    "        #writing positions with VF < 0.002 to one row\n",
    "        abbr1 = df_vcf_C_T[(df_vcf_C_T['VF']<0.002) & (df_vcf_C_T['pos_in_loop']=='-')]\n",
    "        abbr2 = df_vcf_C_T[(df_vcf_C_T['VF']<0.002) & (df_vcf_C_T['pos_in_loop']=='+')]\n",
    "        if len(abbr1) != 0:\n",
    "            df_vcf_C_T_abbr.loc[len(df_vcf_C_T_abbr.index)] = ['1*', 0.002, '-', '-']\n",
    "            flag = True\n",
    "            abbr_text = '1* - ' + str(len(abbr1)) + ' positions not in loops with VF<'+str(0.002)\n",
    "        if len(abbr2) != 0:\n",
    "            df_vcf_C_T_abbr.loc[len(df_vcf_C_T_abbr.index)] = ['2*', 0.002, '-', '+']\n",
    "            new_text = '\\n2* - ' + str(len(abbr2)) + ' positions in loops with VF<'+str(0.002)\n",
    "            abbr_text += new_text\n",
    "        df_vcf_C_T = df_vcf_C_T_abbr.copy()\n",
    "    figure_size = 10\n",
    "    if len(df_vcf_C_T) < 20:\n",
    "        figure_size = 4\n",
    "    \n",
    "    cols = [colors[1] if x == '+' else colors[0] for x in df_vcf_C_T.pos_in_loop]\n",
    "    \n",
    "    f, ax = plt.subplots(figsize=(figure_size, 4))\n",
    "    sns.barplot(x=df_vcf_C_T.index, y=df_vcf_C_T.VF, palette=cols)\n",
    "    f.suptitle('C>T mutations in '+ID+'\\naligned using '+align_method)\n",
    "    ax.set_xticklabels(df_vcf_C_T.POS, rotation = 90)\n",
    "    plt.xticks(fontsize = 4) \n",
    "    ax.set(xlabel='Position in the genome')\n",
    "    \n",
    "    if flag == True:\n",
    "        print(abbr_text)\n",
    "        #add text below figure\n",
    "        plt.subplots_adjust(bottom=0.2)\n",
    "        # Place text outside below the plot\n",
    "        plt.text(1, -0.2, abbr_text, ha='left', fontsize=\"6\")\n",
    "    plt.tight_layout()\n",
    "    plt.close(f)\n",
    "    return(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53e9613b",
   "metadata": {},
   "outputs": [],
   "source": [
    "method = 'BWA'\n",
    "for gse_id in gse_id_list:\n",
    "    \n",
    "    ##get list of srr for gse_id\n",
    "    srr_acc_file_txt = \"SRR_Acc_List_\"+gse_id+\".txt\"\n",
    "    df_srr_acc = pd.read_csv(srr_acc_file_txt, names=['acc'])\n",
    "    srr_list = df_srr_acc.acc.to_list()\n",
    "    \n",
    "    ##path for directory with vcf files for gse_id\n",
    "    directory = 'bwa/vcf_'+gse_id + '/'\n",
    "    for srr_id in srr_list:\n",
    "        print(srr_id, end=', ')\n",
    "        vcf = directory + srr_id + '_sorted.vcf'\n",
    "        plot = apobec_mutagenesis(vcf, srr_id, method)\n",
    "        if type(plot) == str:\n",
    "            print(plot)\n",
    "        else:\n",
    "            plot.savefig(directory+'/pictures/'+srr_id+'_CT_mutations_'+method+'.png', dpi=800)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
