{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "813ae1a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import itertools\n",
    "from Bio import SeqIO\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.patches import Rectangle\n",
    "import seaborn as sns\n",
    "import math"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fdfa2fc6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#input data\n",
    "##list of GSE studies\n",
    "gse_id_list = ['GSE145926', 'GSE154244', 'GSE150316']\n",
    "\n",
    "##min for variant frequency\n",
    "min_VF = 0.001\n",
    "\n",
    "##read genome SARS-CoV-2 file\n",
    "##to find position in genome: str(genome_dict['NC_045512.2'].seq)\n",
    "genome_file = open('SARS_CoV_2.fasta')\n",
    "genome_dict = SeqIO.to_dict(SeqIO.parse(genome_file, \"fasta\"))\n",
    "\n",
    "##palette for pictures\n",
    "colors_list = ['#C6878F', '#3D405B', '#81B29A', '#E07A5F', '#5F797B', '#F2CC8F']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2379bcd3",
   "metadata": {},
   "outputs": [],
   "source": [
    "#function to rounding up top position\n",
    "def round_up(n, decimals=0):\n",
    "    multiplier = 10**decimals\n",
    "    return math.ceil(n * multiplier) / multiplier\n",
    "\n",
    "def get_plot_vcf(vcf_file, min_vf=min_VF, genome=str(genome_dict['NC_045512.2'].seq), colors=colors_list):\n",
    "    \n",
    "    ##read vcf file\n",
    "    SRR_ID = vcf_file.split('/')[-1].split('_')[0]\n",
    "    ##skip description in vcf file\n",
    "    with open(vcf_file, 'r') as f:\n",
    "        reader=f.readlines()\n",
    "        row = 0\n",
    "        while reader[row].startswith('##') == True:\n",
    "            row += 1\n",
    "    df_vcf = pd.read_csv(vcf_file, skiprows=row, sep='\\t')\n",
    "    \n",
    "    ##make dataframe from vcf file\n",
    "    ###list with all allele variants of 3 nucleotides\n",
    "    A = ['A'.join(i) for i in itertools.product('ATCG', repeat=2)]\n",
    "    T = ['T'.join(i) for i in itertools.product('ATCG', repeat=2)]\n",
    "    C = ['C'.join(i) for i in itertools.product('ATCG', repeat=2)]\n",
    "    G = ['G'.join(i) for i in itertools.product('ATCG', repeat=2)]\n",
    "    allele_list = A + T + C + G\n",
    "\n",
    "    ###dataframe for counting reads number for mutations\n",
    "    df_Nreads = pd.DataFrame(columns=['ref_nucl', 'allele', 'A', 'T', 'C', 'G'])\n",
    "    df_Nreads['ref_nucl'] = ['A']*16 + ['T']*16 + ['C']*16 + ['G']*16\n",
    "    df_Nreads['allele'] = allele_list\n",
    "    df_Nreads = df_Nreads.set_index('allele')\n",
    "    df_Nreads.fillna(0, inplace=True)\n",
    "    ###replace cells with no mutation (like A>A) with NaN\n",
    "    for nucl in ['A', 'T', 'C', 'G']:\n",
    "        df_Nreads.loc[df_Nreads['ref_nucl'] == nucl, nucl] = np.nan\n",
    "    \n",
    "    ###dataframe for counting allele's frequencies for mutations\n",
    "    df_freq = pd.DataFrame(columns=['ref_nucl', 'allele', 'A', 'T', 'C', 'G'])\n",
    "    df_freq['ref_nucl'] = ['A']*16 + ['T']*16 + ['C']*16 + ['G']*16\n",
    "    df_freq['allele'] = allele_list\n",
    "    df_freq = df_freq.set_index('allele')\n",
    "    df_freq.fillna(0, inplace=True)\n",
    "    ###replace cells with no mutation (like A>A) with NaN\n",
    "    for nucl in ['A', 'T', 'C', 'G']:\n",
    "        df_freq.loc[df_freq['ref_nucl'] == nucl, nucl] = np.nan\n",
    "        \n",
    "\n",
    "    for row in range(0, len(df_vcf)):\n",
    "        mutation_pos = df_vcf.loc[row, 'POS'] - 1\n",
    "        allele = genome[mutation_pos-1:mutation_pos+2]\n",
    "        alt_nucl = df_vcf.loc[row, 'ALT']\n",
    "        if (len(alt_nucl) != 1) or (len(df_vcf.loc[row, 'REF']) != 1) or (len(allele) != 3):\n",
    "            continue\n",
    "\n",
    "        ###Allele Depth (AD) in vcf file: first number - number of reads with ref nucl, second - with alt nucl (Nreads_alt)\n",
    "        FORMAT = df_vcf.iloc[row, -2].split(':')\n",
    "        AD_ind = FORMAT.index('AD')\n",
    "        VF_ind = FORMAT.index('VF')\n",
    "        AD = df_vcf.iloc[row, -1].split(':')[AD_ind]\n",
    "        VF = df_vcf.iloc[row, -1].split(':')[VF_ind]\n",
    "        Nreads_alt = int(AD.split(',')[1])\n",
    "        freq = float(VF)\n",
    "        if freq >= min_vf:\n",
    "            df_Nreads.at[allele, alt_nucl] += Nreads_alt\n",
    "            df_freq.at[allele, alt_nucl] += freq\n",
    "        \n",
    "    \n",
    "    df_Nreads.reset_index(inplace=True)\n",
    "    df_Nreads = df_Nreads.set_index(['allele'])\n",
    "    df_Nreads = df_Nreads[['A', 'T', 'C', 'G']].stack()\n",
    "    df_Nreads = df_Nreads.to_frame().reset_index()\n",
    "    df_Nreads = df_Nreads.dropna()\n",
    "    df_Nreads = df_Nreads.rename(columns={\"level_1\": \"ALT\", 0:'Nreads'})\n",
    "    df_Nreads['REF'] = df_Nreads['allele'].apply(lambda x: x[1])\n",
    "    df_Nreads['mutation'] = df_Nreads['REF'] + '>' + df_Nreads['ALT']\n",
    "    df_Nreads = df_Nreads[['allele', 'REF', 'ALT', 'mutation', 'Nreads']]\n",
    "    df_Nreads = df_Nreads.sort_values(by=['mutation', 'allele'])\n",
    "    df_Nreads = df_Nreads.reset_index(drop=True)\n",
    "    \n",
    "    df_freq.reset_index(inplace=True)\n",
    "    df_freq = df_freq.set_index(['allele'])\n",
    "    df_freq = df_freq[['A', 'T', 'C', 'G']].stack()\n",
    "    df_freq = df_freq.to_frame().reset_index()\n",
    "    df_freq = df_freq.dropna()\n",
    "    df_freq = df_freq.rename(columns={\"level_1\": \"ALT\", 0:'freq'})\n",
    "    df_freq['REF'] = df_freq['allele'].apply(lambda x: x[1])\n",
    "    df_freq['mutation'] = df_freq['REF'] + '>' + df_freq['ALT']\n",
    "    df_freq = df_freq[['allele', 'REF', 'ALT', 'mutation', 'freq']]\n",
    "    df_freq = df_freq.sort_values(by=['mutation', 'allele'])\n",
    "    df_freq = df_freq.reset_index(drop=True)\n",
    "    \n",
    "\n",
    "    #make picture\n",
    "    custom_palette = []\n",
    "    for x in colors:\n",
    "        custom_palette.extend([x]*16)\n",
    "    ##divide dataframe into two parts\n",
    "    df1_Nreads = df_Nreads[df_Nreads['REF'].isin(['A', 'C'])]\n",
    "    df2_Nreads = df_Nreads[df_Nreads['REF'].isin(['G', 'T'])]\n",
    "    df1_freq = df_freq[df_freq['REF'].isin(['A', 'C'])]\n",
    "    df2_freq = df_freq[df_freq['REF'].isin(['G', 'T'])]\n",
    "\n",
    "    mutation_list1 = df1_Nreads.mutation.to_list()\n",
    "    mutation_list1 = list(set(mutation_list1))\n",
    "    mutation_list1.sort()\n",
    "    mutation_list2 = df2_Nreads.mutation.to_list()\n",
    "    mutation_list2 = list(set(mutation_list2))\n",
    "    mutation_list2.sort()\n",
    "\n",
    "    ###picture for number of reads\n",
    "    fig1 = plt.figure(figsize=(16, 7))\n",
    "    ax11 = plt.subplot2grid(shape=(2, 1), loc=(0, 0))\n",
    "    ax12 = plt.subplot2grid(shape=(2, 1), loc=(1, 0))\n",
    "\n",
    "    ax11 = sns.barplot(x=df1_Nreads.index, y=df1_Nreads.Nreads, palette=custom_palette, ax=ax11)\n",
    "    ax12 = sns.barplot(x=df2_Nreads.index, y=df2_Nreads.Nreads, palette=custom_palette, ax=ax12)\n",
    "    ax11.set_xticklabels(df1_Nreads.allele, rotation=90, horizontalalignment='center')\n",
    "    ax12.set_xticklabels(df2_Nreads.allele, rotation=90, horizontalalignment='center')\n",
    "\n",
    "    ####change lim of picture\n",
    "    top_position = int(df_Nreads.Nreads.max())\n",
    "    round_pos = 0\n",
    "    while top_position >= 10:\n",
    "        top_position = int(top_position / 10)\n",
    "        round_pos += 1\n",
    "    new_top_position = round_up(df_Nreads.Nreads.max(), -round_pos)\n",
    "\n",
    "    ####rectangle must place 10% of space\n",
    "    part_ten = int(new_top_position/10)\n",
    "    ax11.set_ylim(0, new_top_position+part_ten)\n",
    "    ax12.set_ylim(0, new_top_position+part_ten)\n",
    "\n",
    "    ####add rectangles\n",
    "    for i in range(0, len(colors)):\n",
    "        ax11.add_patch(Rectangle((i*16, new_top_position), 16, part_ten, facecolor = colors[i]))\n",
    "        ax11.text(i*16+8, new_top_position+int(part_ten/2), mutation_list1[i], horizontalalignment='center', verticalalignment='center', size='x-large', color='white', weight='semibold')\n",
    "        ax12.add_patch(Rectangle((i*16, new_top_position), 16, part_ten, facecolor = colors[i]))\n",
    "        ax12.text(i*16+8, new_top_position+int(part_ten/2), mutation_list2[i], horizontalalignment='center', verticalalignment='center', size='x-large', color='white', weight='semibold')\n",
    "\n",
    "    ####save picture\n",
    "    fig1.suptitle('Number of reads by mutation type for sample '+SRR_ID, fontsize=18)\n",
    "    plt.tight_layout()\n",
    "    plt.close(fig1)\n",
    "\n",
    "\n",
    "    ###picture for freq\n",
    "    fig2 = plt.figure(figsize=(16, 7))\n",
    "    ax21 = plt.subplot2grid(shape=(2, 1), loc=(0, 0))\n",
    "    ax22 = plt.subplot2grid(shape=(2, 1), loc=(1, 0))\n",
    "\n",
    "    ax21 = sns.barplot(x=df1_freq.index, y=df1_freq.freq, palette=custom_palette, ax=ax21)\n",
    "    ax22 = sns.barplot(x=df2_freq.index, y=df2_freq.freq, palette=custom_palette, ax=ax22)\n",
    "    ax21.set_xticklabels(df1_freq.allele, rotation=90, horizontalalignment='center')\n",
    "    ax22.set_xticklabels(df2_freq.allele, rotation=90, horizontalalignment='center')\n",
    "\n",
    "    ####change lim of picture\n",
    "    top_position = df_freq.freq.max()\n",
    "    if top_position >= 1:\n",
    "        new_top_position = top_position + 0.1\n",
    "    else:\n",
    "        round_pos = 0\n",
    "        while top_position < 1:\n",
    "            top_position = top_position * 10\n",
    "            round_pos += 1\n",
    "        new_top_position = round_up(df_freq.freq.max(), round_pos)\n",
    "        \n",
    "\n",
    "    ####rectangle must place 10% of space\n",
    "    part_ten = new_top_position/10\n",
    "    ax21.set_ylim(0, new_top_position+part_ten)\n",
    "    ax22.set_ylim(0, new_top_position+part_ten)\n",
    "    \n",
    "\n",
    "    ####add rectangles\n",
    "    for i in range(0, len(colors)):\n",
    "        ax21.add_patch(Rectangle((i*16, new_top_position), 16, part_ten, facecolor = colors[i]))\n",
    "        ax21.text(i*16+8, new_top_position+part_ten/2, mutation_list1[i], horizontalalignment='center', verticalalignment='center', size='x-large', color='white', weight='semibold')\n",
    "        ax22.add_patch(Rectangle((i*16, new_top_position), 16, part_ten, facecolor = colors[i]))\n",
    "        ax22.text(i*16+8, new_top_position+part_ten/2, mutation_list2[i], horizontalalignment='center', verticalalignment='center', size='x-large', color='white', weight='semibold')\n",
    "\n",
    "    fig2.suptitle('Allele frequency by mutation type for sample '+SRR_ID, fontsize=18)\n",
    "    plt.tight_layout()\n",
    "    plt.close(fig2)\n",
    "    \n",
    "    return [fig1, fig2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "76230db9",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig_list = ['Nreads', 'freq']\n",
    "\n",
    "for gse_id in gse_id_list:\n",
    "    \n",
    "    ##get list of srr for gse_id\n",
    "    srr_acc_file_txt = \"SRR_Acc_List_\"+gse_id+\".txt\"\n",
    "    df_srr_acc = pd.read_csv(srr_acc_file_txt, names=['acc'])\n",
    "    srr_list = df_srr_acc.acc.to_list()\n",
    "    \n",
    "    ##path for directory with vcf files for gse_id\n",
    "    directory = 'bwa/vcf_'+gse_id\n",
    "    for srr_id in srr_list:\n",
    "        print(srr_id, end=', ')\n",
    "        vcf = directory + '/' + srr_id + '_sorted.vcf'\n",
    "        plot_list = get_plot_vcf(vcf)\n",
    "        for i in range(0, len(plot_list)):\n",
    "            plot_list[i].savefig(directory+'/pictures/'+srr_id+'_'+fig_list[i]+'.png')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
